### PR TITLE
Simplify exception handling; cache stream loader initialization result

### PR DESCRIPTION
**Pull request by johnny-schmidt**
### Simplify exception handling

the old algorithm was too convoluted

now:
  * failure of any task triggers a "for each stream, mark failed and run CloseStream(failure)"
  * then fail the sync and run Teardown(failure)
  * then kill the scopes
 
This allows us to get rid of the task `Level`s entirely.

### Stream Loader initialization must return a result

the old way was only to initialize on success and cache the streamloader

so any task that needed a stream loader (process records, process batch) would hang forever waiting for the failed one

now we catch initialization exceptions and cache the result

anything that tries to look up a stream loader that failed to initialize gets the exception

### Testing

I could not reproduce the hanging failure in the cloud, but I verified that some ITs hang forever on `StreamLoader::start` failures w/o this fix.

NOTE: Why isn't this two PRs? Because for some reason just the stream loader fix triggered a weird unrelated micronaut error (can't inject something totally unrelated that obviously exists?) that I didn't feel like debugging.

### CODE REVIEW

File: airbyte-cdk/bulk/core/load/src/test/kotlin/io/airbyte/cdk/load/task/DestinationTaskLauncherTest.kt [hunk 1]
```diff
@@ -544,6 +538,42 @@ class DestinationTaskLauncherTest<T> where T : LeveledTask, T : ScopedTask {
             taskLauncher.run()
             Assertions.assertTrue(mockScopeProvider.didKill)
         }
-        mockExceptionHandler.callbacks.receive().invoke()
+        taskLauncher.handleTeardownComplete(success = false)
+    }
+
+    @Test
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/test/kotlin/io/airbyte/cdk/load/task/DestinationTaskLauncherTest.kt:544]: I found it impossible to get either of these tests to work with mockk. I kept getting bizarre unrelated errors (like failure to mock stuff inside of some method calls regardless of whether I stubbed them). Nor could I find any way to force a failure inside a task in a context that would trigger the callback.

**Comment by edgao** [comment]
@frifriSF59 fyi - this might help with the testInterruptedTruncateWithoutPriorData thing

(https://github.com/airbytehq/airbyte/pull/48864/files#diff-e40740d64bd693a3f03a0c1b1e0ab258253cffc4121b670ca04a03f9c040a4f7R63 - the connector just stops progressing at some point. Didn't dig into where exactly that happens though, i.e. might not have been during streamloader start)

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/internal/SpillToDiskTask.kt [hunk 2]
```diff
@@ -70,7 +69,6 @@ class DefaultSpillToDiskTask(
                         reserved.use {
                             when (val wrapped = it.value) {
                                 is StreamRecordEvent -> {
-                                    // once we have received a record for the stream, consider the
```
Comment (tryangul) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/internal/SpillToDiskTask.kt:73]: Nit: Undelete

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/DestinationTaskLauncher.kt [hunk 3]
```diff
@@ -132,14 +141,38 @@ class DefaultDestinationTaskLauncher(
     private val succeeded = Channel<Boolean>(Channel.UNLIMITED)
 
     private val teardownIsEnqueued = AtomicBoolean(false)
+    private val failSyncIsEnqueued = AtomicBoolean(false)
 
-    private suspend fun enqueue(task: LeveledTask) {
-        taskScopeProvider.launch(exceptionHandler.withExceptionHandling(task))
+    inner class TaskWrapper(
```
Comment (tryangul) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/DestinationTaskLauncher.kt:146]: Nit: should this be named something that indicates its purpose is to do error handling?

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/DestinationTaskLauncher.kt [hunk 4]
```diff
@@ -132,14 +141,38 @@ class DefaultDestinationTaskLauncher(
...
 
-    override suspend fun run() {
-        exceptionHandler.setCallback { succeeded.send(false) }
+    inner class NoopWrapper(
+        override val innerTask: ScopedTask,
+    ) : WrappedTask<ScopedTask> {
+        override suspend fun execute() {
+            innerTask.execute()
+        }
+    }
```
Comment (tryangul) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/DestinationTaskLauncher.kt:168]: If `WrappedTask` implements `Task`, why do we need this?

**Review by tryangul — APPROVED** [review]

**Pull request merged**